File: ScanbotSDKDemo/BlurrinessEstimatorDemoViewController.m
//
//  BlurrinessEstimatorDemoViewController.m
//  SBSDK Internal Demo
//
//  Created by Sebastian Husche on 07.07.20.
//  Copyright © 2020 doo GmbH. All rights reserved.
//

#import "BlurrinessEstimatorDemoViewController.h"
@import ScanbotSDK;

@interface BlurrinessEstimatorDemoViewController() <SBSDKScannerViewControllerDelegate, UINavigationControllerDelegate,
UIImagePickerControllerDelegate>

@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (strong, nonatomic) SBSDKScannerViewController *scannerViewController;
@property (strong, nonatomic) SBSDKBlurrinessEstimator *estimator;
@end

@implementation BlurrinessEstimatorDemoViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.scannerViewController = [[SBSDKScannerViewController alloc] initWithParentViewController:self
                                                                                       parentView:self.containerView
                                                                                     imageStorage:nil
                                                                            enableQRCodeDetection:NO];
    self.estimator = [[SBSDKBlurrinessEstimator alloc] init];
    self.scannerViewController.delegate = self;
    self.scannerViewController.detectionStatusHidden = YES;
}

- (IBAction)selectImageButtonTapped:(id)sender {
    [self darkenScreen];
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.modalPresentationStyle = UIModalPresentationFullScreen;
    [self presentViewController:picker animated:YES completion:nil];
}

- (void)estimateAndShowResultsWithImage:(UIImage *)image {
    double result = [self.estimator estimateImageBlurriness:image];
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        [weakSelf showResult:result];
    });
}

- (void)darkenScreen {
    [UIView animateWithDuration:0.2 animations:^{
        self.scannerViewController.HUDView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.85];
    }];
}

- (void)lightenScreen {
    [UIView animateWithDuration:0.2 animations:^{
        self.scannerViewController.HUDView.backgroundColor = [UIColor clearColor];
    }];
}


- (void)showResult:(double)result {
    NSString *resultString = [NSString stringWithFormat:@"Blurriness = %0.0f%%", result * 100.0f];
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Blurriness Estimation"
                                                                   message:resultString
                                                           preferredStyle:UIAlertControllerStyleAlert];
    __weak typeof(self) weakSelf = self;
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                       style:UIAlertActionStyleDefault
                                                     handler:^(UIAlertAction * action) {
        [weakSelf lightenScreen];
    }];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
 }


// MARK: - Picker
- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    __weak typeof(self) weakSelf = self;
    [self dismissViewControllerAnimated:YES completion:^{
        [weakSelf lightenScreen];
    }];
}

- (void)imagePickerController:(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    UIImage *image = (UIImage *)info[UIImagePickerControllerOriginalImage];
    __weak typeof(self) weakSelf = self;
    [self dismissViewControllerAnimated:YES completion:^{
        [weakSelf estimateAndShowResultsWithImage:image];
    }];
}

// MARK: - Scanner delegate
- (BOOL)scannerControllerShouldAnalyseVideoFrame:(SBSDKScannerViewController *)controller {
    return NO;
}

- (void)scannerControllerWillCaptureStillImage:(SBSDKScannerViewController *)controller {
    [self darkenScreen];
}

- (void)scannerController:(nonnull SBSDKScannerViewController *)controller didCaptureImage:(nonnull UIImage *)image {
    [self estimateAndShowResultsWithImage:image];
}

- (void)scannerController:(SBSDKScannerViewController *)controller didFailCapturingImage:(NSError *)error {
    [self lightenScreen];
}
@end

File: ScanbotSDKDemo/Utilities/BlurrinessEstimatorDemoViewController.swift
//
//  BlurrinessEstimatorDemoViewController.swift
//  ScanbotSDKDemo
//
//  Created by Danil Voitenko on 24.11.21.
//  Copyright © 2021 doo GmbH. All rights reserved.
//

import UIKit
import ScanbotSDK

final class BlurrinessEstimatorDemoViewController: UIViewController {
    @IBOutlet private var containerView: UIView?
    private var scannerViewController: SBSDKDocumentScannerViewController?
    private var estimator: SBSDKBlurrinessEstimator?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        scannerViewController = SBSDKDocumentScannerViewController(parentViewController: self,
                                                                   parentView: containerView,
                                                                   delegate: self)
        estimator = SBSDKBlurrinessEstimator()
        scannerViewController?.delegate = self
        scannerViewController?.isAutoSnappingEnabled = false
        scannerViewController?.hideDetectionStatusLabel = true
    }
    
    @IBAction private func selectImageButtonDidPress(_ sender: Any) {
        darkenScreen()
        let picker = UIImagePickerController()
        picker.delegate = self
        picker.modalPresentationStyle = .fullScreen
        present(picker, animated: true, completion: nil)
    }
    
    private func estimateAndShowResults(from image: UIImage) {
        if let result = estimator?.estimateImageBlurriness(image) {
            DispatchQueue.main.async { [weak self] in
                self?.show(result: result)
            }
        }
    }
    
    private func darkenScreen() {
        UIView.animate(withDuration: 0.2) { [weak self] in
            self?.scannerViewController?.view.backgroundColor = UIColor.black.withAlphaComponent(0.8)
            
        }
    }
    
    private func lightenScreen() {
        UIView.animate(withDuration: 0.2) { [weak self] in
            self?.scannerViewController?.view.backgroundColor = UIColor.clear
            
        }
    }
    
    private func show(result: Double) {
        let resultString = "Bluriness = \(result * 100)"
        let alert = UIAlertController(title: "Bluriness Estimation",
                                      message: resultString,
                                      preferredStyle: .alert)
        let okAction = UIAlertAction(title: "OK",
                                     style: .default) { _ in
            alert.presentedViewController?.dismiss(animated: true)
        }
        alert.addAction(okAction)
        present(alert, animated: true, completion: nil)
    }
    
    
}

extension BlurrinessEstimatorDemoViewController: SBSDKDocumentScannerViewControllerDelegate {
    func documentScannerViewController(_ controller: SBSDKDocumentScannerViewController,
                                       didSnapDocumentImage documentImage: UIImage,
                                       on originalImage: UIImage,
                                       with result: SBSDKDocumentDetectorResult,
                                       autoSnapped: Bool) {
        estimateAndShowResults(from: originalImage)
    }
}

extension BlurrinessEstimatorDemoViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController,
                               didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        if let image = info[UIImagePickerController.InfoKey.originalImage] as? UIImage {
            dismiss(animated: true) { [weak self] in
                self?.estimateAndShowResults(from: image)
            }
        } else {
            dismiss(animated: true, completion: nil)
        }
    }
}
